vm/src/builtins/int.rs

```diff
@@ -78,6 +78,18 @@ impl PyValue for PyInt {
     fn into_object(self, vm: &VirtualMachine) -> PyObjectRef {
         vm.ctx.new_int(self.value)
     }
+
+    fn special_retrieve(vm: &VirtualMachine, obj: PyObjectRef) -> Option<PyResult<PyRef<Self>>> {
+        vm.get_method(obj, "__index__").map(|index| {
+            // TODO: returning strict subclasses of int in __index__ is deprecated
+            vm.invoke(&index?, ())?.downcast().map_err(|bad| {
+                vm.new_type_error(format!(
+                    "__index__ returned non-int (type {})",
+                    bad.class().name
+                ))
+            })
+        })
```

Comment:
You can just call `vm.to_index()` here, we already have this logic implemented.

Reply:
I added it as `to_index_opt()` so I could return `None` in the default implementation, is that OK?

Reply:
Well I think it makes sense to always return `Some`, since `to_index` will have a better error message.

Reply:
Ah, I was using the `pyref_type_error` for all circumstances.

So, have the default implementation of `special_retrieve` return the current `pyref_type_error`, and pass `to_index` unchanged from the int `special_retrieve`?

Reply:
Like this?